src/mongodb-migrations.ts

```diff
@@ -0,0 +1,377 @@
+import fs from 'fs';
+import path from 'path';
+import Promise from 'bluebird';
+import _ from 'lodash';
+import { Db, Collection, MongoClient, Document, WithId, AnyError, Callback, InsertOneResult, DeleteResult } from 'mongodb-legacy';
+import { repeatString, connect as mongoConnect, normalizeConfig } from './utils';
+import { MongoConfig } from './url-builder';
+import { mkdirp } from 'mkdirp';
+
+// Import migration stub as a require since it uses module.exports
+const migrationStub = require('./migration-stub');
+
+export type LogLevel = 'system' | 'user';
+type LogFunction = ((level: LogLevel, message: string) => void) | null;
+
+const defaultLog = (src: LogLevel, ...args: any[]): void => {
+  const pad = repeatString(' ', src === 'system' ? 4 : 2);
+  console.log(pad, ...args);
+};
+
+type MigrationCallback = (error?: Error) => void;
+
+// A migration function can either:
+// 1. Return a Promise<void>
+// 2. Accept a callback parameter
+type MigrationFunction = 
+  | ((this: MigrationContext) => Promise<void>)
+  | ((this: MigrationContext, done: MigrationCallback) => void);
+
+interface Migration {
+  id: string;
+  up?: MigrationFunction;
+  down?: MigrationFunction;
+}
+
+type MigrationStatus = 'ok' | 'skip' | 'error';
+
+interface MigrationResult {
+  status: MigrationStatus;
+  error?: Error;
+  reason?: string;
+  code?: 'no_up' | 'no_down' | 'already_ran' | 'not_in_recent_migrate';
+}
+
+interface MigrationResults {
+  [key: string]: MigrationResult;
+}
+
+interface MigrationContext {
+  db: Db;
+  log: (message: string) => void;
+  client: MongoClient;
+}
+
+export class Migrator {
+  private _isDisposed: boolean;
+  private _m: Migration[];
+  private _result: MigrationResults;
+  private _dbReady: Promise<void>;
+  private _db!: Db;
+  private _client!: MongoClient;
+  private _collName: string;
+  private _timeout?: number;
+  private _ranMigrations: { [key: string]: boolean } = {};
+  private _lastDirection?: 'up' | 'down';
+  private log: LogFunction;
+
+  constructor(dbConfig: MongoConfig, logFn?: LogFunction) {
+    // this will throw in case of invalid values
+    dbConfig = normalizeConfig(dbConfig);
+
+    this._isDisposed = false;
+    this._m = [];
+    this._result = {};
+
+    this._dbReady = (Promise.fromCallback as any)((cb: (error?: Error) => void) => {
+      mongoConnect(dbConfig, cb);
+    }).then((client: MongoClient) => {
+      this._client = client;
+      this._db = client.db();
+    });
+
+    this._collName = dbConfig.collection!;
+    this._timeout = dbConfig.timeout;
+
+    if (logFn || logFn === null) {
+      this.log = logFn;
+    } else {
+      this.log = defaultLog;
+    }
+  }
+
+  add(m: Migration): void {
+    // m must be an { id, up, down } object
+    this._m.push(m);
+  }
+
+  bulkAdd(array: Migration[]): void {
+    // array must be an Array of { id, up, down } objects
+    this._m = this._m.concat(array);
+  }
+
+  private _coll(): Collection<Document> {
```

Comment:
The underscores aren't necessary. Could probably benefit from better naming convention as well.